File: ros2_ws/src/multimedia_cpp/src/si470x/si470x_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/int32.hpp>

#include "multimedia_cpp/si470x/si470x.hpp"

class Si470xNode : public rclcpp::Node
{
public:
    Si470xNode()
    : Node("si470x_node"),
      radio_(
        declare_parameter<std::string>("i2c_device", "/dev/i2c-1"),
        declare_parameter<int>("i2c_address", 0x10),
        declare_parameter<int>("reset_gpio", 27))
    {
        // Frequency range parameters (in MHz)
        min_freq_mhz_ = declare_parameter<double>("min_freq_mhz", 87.5).get<double>();
        max_freq_mhz_ = declare_parameter<double>("max_freq_mhz", 108.0).get<double>();

        if (!radio_.init())
        {
            RCLCPP_FATAL(get_logger(), "Failed to initialize Si470x");
            throw std::runtime_error("Si470x init failed");
        }

        // Subscribe to final frequency in 0.1 MHz units (e.g. 995 = 99.5 MHz)
        sub_ = create_subscription<std_msgs::msg::Int32>(
            "/current/frequency_int",
            10,
            std::bind(&Si470xNode::on_frequency, this, std::placeholders::_1));

        RCLCPP_INFO(
            get_logger(),
            "Si470x node started. Listening to /current/frequency_int (0.1 MHz units). "
            "Range: [%.1f, %.1f] MHz",
            min_freq_mhz_, max_freq_mhz_);
    }

private:
    void on_frequency(const std_msgs::msg::Int32::SharedPtr msg)
    {
        int freq10 = msg->data;  // 0.1 MHz units

        // Clamp based on configured MHz range
        int min10 = static_cast<int>(min_freq_mhz_ * 10.0 + 0.5);
        int max10 = static_cast<int>(max_freq_mhz_ * 10.0 + 0.5);

        if (freq10 < min10) freq10 = min10;
        if (freq10 > max10) freq10 = max10;

        float freq_mhz = freq10 / 10.0f;

        RCLCPP_INFO(
            get_logger(),
            "Tuning Si470x to %.1f MHz (freq10=%d)",
            freq_mhz, freq10);

        if (!radio_.set_frequency_10k(freq10))
        {
            RCLCPP_ERROR(
                get_logger(),
                "Failed to tune Si470x to %.1f MHz (freq10=%d)",
                freq_mhz, freq10);
        }
    }

    Si470x radio_;
    rclcpp::Subscription<std_msgs::msg::Int32>::SharedPtr sub_;

    double min_freq_mhz_;
    double max_freq_mhz_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Si470xNode>());
    rclcpp::shutdown();
    return 0;
}

File: ros2_ws/src/multimedia_python/multimedia_python/si470x_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from std_msgs.msg import Int32, Float32

from si4703Library import si4703Radio


class Si470xNode(Node):

    def __init__(self):
        super().__init__('si470x_node')

        addr = self.declare_parameter('i2c_address', 0x10).value
        rst = self.declare_parameter('reset_pin', 5).value
        irq = self.declare_parameter('irq_pin', -1).value

        if irq == -1:
            self.radio = si4703Radio(addr, rst)
        else:
            self.radio = si4703Radio(addr, rst, irq)

        self.radio.si4703Init()

        self.create_subscription(
            Int32,   '/current/frequency_int',   self.cb_freq, 10)
        self.create_subscription(
            Float32, '/current/volume_float32',  self.cb_vol,  10)

        self.get_logger().info("Si470x Node started.")

    def cb_freq(self, msg):
        ch = int(msg.data)
        try:
            self.radio.si4703SetChannel(ch)
            self.get_logger().info(f"Set freq -> {ch/10:.1f} MHz")
        except Exception as e:
            self.get_logger().error(str(e))

    def cb_vol(self, msg):
        vol_norm = max(0.0, min(1.0, float(msg.data)))
        hw = int(round(vol_norm * 15))
        try:
            self.radio.si4703SetVolume(hw)
            self.get_logger().info(f"Set volume -> {hw}")
        except Exception as e:
            self.get_logger().error(str(e))


def main(args=None):
    rclpy.init(args=args)
    node = Si470xNode()
    rclpy.spin(node)
    rclpy.shutdown()
